core/domain/exp_services.py

```diff
@@ -1707,6 +1707,350 @@ def is_voiceover_change_list(change_list):
+        # according to the state and property name.
+
+        added_state_names = []
+        deleted_state_names = []
+        new_to_old_state_names = {}
+        changed_properties = {}
+
+        old_version = exp_fetchers.get_exploration_by_id(
+            exp_id, version=version)
+        new_version = exp_fetchers.get_exploration_by_id(
+            exp_id, version=latest_version)
+        for change in composite_change_list:
+            if change.cmd == exp_domain.CMD_ADD_STATE:
+                added_state_names.append(change.state_name)
+            elif change.cmd == exp_domain.CMD_DELETE_STATE:
+                state_name = change.state_name
+                if state_name in added_state_names:
+                    added_state_names.remove(state_name)
+                else:
+                    original_state_name = state_name
+                    if original_state_name in new_to_old_state_names:
+                        original_state_name = new_to_old_state_names.pop(
+                            original_state_name)
+                    deleted_state_names.append(original_state_name)
+            elif change.cmd == exp_domain.CMD_RENAME_STATE:
+                old_state_name = change.old_state_name
+                new_state_name = change.new_state_name
+                if old_state_name in added_state_names:
+                    added_state_names.remove(old_state_name)
+                    added_state_names.append(new_state_name)
+                elif old_state_name in new_to_old_state_names:
+                    new_to_old_state_names[new_state_name] = (
+                        new_to_old_state_names.pop(old_state_name))
+                else:
+                    new_to_old_state_names[new_state_name] = old_state_name
+
+            # A condition to store the name of the properties changed
+            # in changed_properties dict.
+            elif change.cmd == exp_domain.CMD_EDIT_STATE_PROPERTY:
+                state_name = change.state_name
+                if state_name in new_to_old_state_names:
+                    state_name = new_to_old_state_names[change.state_name]
+                if state_name in changed_properties:
+                    if (change.property_name not in
+                            changed_properties[state_name]):
+                        changed_properties[state_name].append(
+                            change.property_name)
+                else:
+                    changed_properties[state_name] = [change.property_name]
+
+        old_to_new_state_names = {
+            value: key for key, value in new_to_old_state_names.items()
+        }
+
+        if len(added_state_names) > 0 or len(deleted_state_names) > 0:
+            # Here we will send the changelist, version, latest_version,
+            # and exploration to the admin, so that the conditions
+            # can we reviewed.
+            return False
+
+        changes_are_mergeable = False
+        state_names_of_renamed_states = {}
```

Comment:
Done.

Reply:
"state_names_of_renamed_states" renamed states sounds like it will be a list of state names. If it has to be a dict, what is the structure of the dict (i.e. what are the keys and what are the values)? Please explain in comments.

Reply:
Done.